src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -125,5 +139,168 @@ public override bool IsInvalid
+                // else and we would block trying to access it.  Other errors, such as ENOTSUP (locking isn't supported) or
+                // EACCES (the file system doesn't allow us to lock), will only hamper FileStream's usage without providing value,
+                // given again that this is only advisory / best-effort.
+                Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                if (errorInfo.Error == Interop.Error.EWOULDBLOCK)
+                {
+                    throw Interop.GetExceptionForIoErrno(errorInfo, path, isDirectory: false);
+                }
+            }
+
+            // These provide hints around how the file will be accessed.  Specifying both RandomAccess
+            // and Sequential together doesn't make sense as they are two competing options on the same spectrum,
+            // so if both are specified, we prefer RandomAccess (behavior on Windows is unspecified if both are provided).
+            Interop.Sys.FileAdvice fadv =
+                (options & FileOptions.RandomAccess) != 0 ? Interop.Sys.FileAdvice.POSIX_FADV_RANDOM :
+                (options & FileOptions.SequentialScan) != 0 ? Interop.Sys.FileAdvice.POSIX_FADV_SEQUENTIAL :
+                0;
+            if (fadv != 0)
+            {
+                FileStreamHelpers.CheckFileCall(Interop.Sys.PosixFAdvise(this, 0, 0, fadv), path,
+                    ignoreNotSupported: true); // just a hint.
+            }
+
+            if (mode == FileMode.Create || mode == FileMode.Truncate)
+            {
+                // Truncate the file now if the file mode requires it. This ensures that the file only will be truncated
+                // if opened successfully.
+                if (Interop.Sys.FTruncate(this, 0) < 0)
+                {
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                    if (errorInfo.Error != Interop.Error.EBADF && errorInfo.Error != Interop.Error.EINVAL)
+                    {
+                        // We know the file descriptor is valid and we know the size argument to FTruncate is correct,
+                        // so if EBADF or EINVAL is returned, it means we're dealing with a special file that can't be
+                        // truncated.  Ignore the error in such cases; in all others, throw.
+                        throw Interop.GetExceptionForIoErrno(errorInfo, path, isDirectory: false);
+                    }
+                }
+            }
+
+            // If preallocationSize has been provided for a creatable and writeable file
+            if (FileStreamHelpers.ShouldPreallocate(preallocationSize, access, mode))
+            {
+                int fallocateResult = Interop.Sys.PosixFAllocate(this, 0, preallocationSize);
+                if (fallocateResult != 0)
+                {
+                    Dispose();
+                    Interop.Sys.Unlink(path!); // remove the file to mimic Windows behaviour (atomic operation)
```

Comment:
We remove the file only when `preallocationSize` is specified because this is the only time when we are using `posix_fallocate`. `posix_fallocate` might allocate the entire available disk space in cases where there is not enough disk space to satisfy the request.

Example: we have 1 GB of free space and the user specifies 2 GB as `preallocationSize`. `posix_fallocate` allocates 1 GB and then fails with `ENOSPC`. The disk is full.

To mimic the Windows behaviour, where `NtCreateFile` just throws and don't allocate any disk space if it can't satisfy the entire request, we close the handle and then unlink it.

Reply:
> To mimic the Windows behaviour, where NtCreateFile just throws and don't allocate any disk space if it can't satisfy the entire request, we close the handle and then unlink it.

In that case NtCreateFile doesn't leave behind an empty file?

Reply:
> In that case NtCreateFile doesn't leave behind an empty file?

It does not and we have a test that ensures that:

https://github.com/dotnet/runtime/blob/8709d595f7105cbf73774ec762266bc56f99cf04/src/libraries/System.IO.FileSystem/tests/FileStream/ctor_options_as.cs#L144-L161

Reply:
Ok.  FWIW, it's strange to me that we now have an inconsistency as to whether the file is there or not based on the argument to preallocationSize.  But if there's really no good way around it, ok. Thanks.